### ISSUE TITLE
opencv dnn cannot accept image with 2 channel

### ISSUE THREAD

**Issue by stereomatchingkiss** [opencv/opencv#9747]
```
input_shape = [None,128,128,2]
features = tf.placeholder(tf.float32, input_shape, name = 'input')
net = tf.layers.conv2d(inputs = features, filters = 64, kernel_size = [3, 3], strides = (2, 2), padding = 'same')    
net = tf.contrib.layers.batch_norm(net, is_training = False)
net = tf.nn.relu(net)
net = tf.reshape(net, [-1, 64 * 64 * 64])
net = tf.layers.dense(inputs = net, units = 8, kernel_initializer = tf.contrib.layers.xavier_initializer(), name = 'regression_output')

global_step = tf.Variable(0, trainable=False)
boundaries = [30000, 60000]
values = [0.005, 0.0005, 0.00005]
learning_rate = tf.train.piecewise_constant(global_step, boundaries, values)

x = tf.placeholder(tf.float32, shape=[None,8])
#apply sqrt to avoid the loss value bloated
loss = tf.sqrt((tf.nn.l2_loss(net - x)))

with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    saver = tf.train.Saver()
    saver.save(sess, 'reshape_final.ckpt')
    tf.train.write_graph(sess.graph.as_graph_def(), "", 'graph_final.pb')
```

After that I freeze graph

```
python3 ~/.keras2/lib/python3.5/site-packages/tensorflow/python/tools/freeze_graph.py --input_graph=graph_final.pb --input_checkpoint=reshape_final.ckpt --output_graph=frozen_graph.pb --output_node_names=regression_output/BiasAdd

python3 ~/.keras2/lib/python3.5/site-packages/tensorflow/python/tools/optimize_for_inference.py --input frozen_graph.pb --output opt_graph.pb --frozen_graph True --input_names input --output_names regression_output/BiasAdd

~/Qt/3rdLibs/tensorflow/bazel-bin/tensorflow/tools/graph_transforms/transform_graph --in_graph=opt_graph.pb --out_graph=fused_graph.pb --inputs=input --outputs=regression_output/BiasAdd --transforms="fold_constants sort_by_execution_order"

```

Then I forward the graph

```
std::string const model("/home/ramsus/Qt/blogCodes2/deep_homography/cnn/tensorflow/fused_graph.pb");

dnn::Net net = dnn::readNetFromTensorflow(model);
if(net.empty()){
      std::cerr<<"Can't load network by using the mode file:"<<std::endl;
      std::cerr<<model<<std::endl;
      return -1;
}

cv::Mat img(128, 128, CV_8UC2);
Mat inputBlob = dnn::blobFromImage(img); //Convert Mat to image batch

```

Error messages:

> OpenCV Error: Assertion failed (nch == 1) in blobFromImages, file /home/ramsus/Qt/3rdLibs/opencv/modules/dnn/src/dnn.cpp, line 155
> /home/ramsus/Qt/3rdLibs/opencv/modules/dnn/src/dnn.cpp:155: error: (-215) nch == 1 in function blobFromImages

By the way, do anyone know how to change is_training of batchnorm to false(I put complete question at [stackoverflow](https://stackoverflow.com/questions/46498332/turn-is-training-of-batchnormtensorflow-to-false))? dnn module cannot load the model if is_training is True, I have to change it back to False and save the model again.

**Comment by stereomatchingkiss** [comment]
I tried to create the input blob by myself

    ```
cv::Mat img(128, 128, CV_32FC2);
    int sz[] = { 1, 2, img.rows, img.cols };
    cv::Mat blob = Mat(4, sz, CV_32F);

    CV_Assert(img.dims == 2);

    cv::Mat mat[2];
    cv::split(img, mat);
    mat[0].copyTo(Mat(img.rows, img.cols, CV_32F, blob.ptr((0, 0))));
    mat[1].copyTo(Mat(img.rows, img.cols, CV_32F, blob.ptr((0, 1))));
    net.setInput(blob, "input"); //set the network input
    Mat result = net.forward(); //compute output
```

This time give me error messages

> tensorflow_model_loader': malloc(): smallbin double linked list corrupted: 0x00000000024d1490 ***

Either I create two dimensions blobs with wrong solution or net.forward do not support 2 channels blob(it works if I change the channel to 1 dimension).

**Comment by stereomatchingkiss** [comment]
I try to split the input images into two parts, from [batch_size,128, 128, 2] to [batch_size*2, 128, 128, 1].
After that I handle the channels separately

```
in1 = tf.layers.conv2d(inputs = features[0:batch_size], filters = 64, kernel_size = [3, 3], strides = (2, 2), padding = 'same')    
in1 = tf.contrib.layers.batch_norm(in1, is_training = is_training)
    
in2 = tf.layers.conv2d(inputs = features[batch_size:batch_size*2], filters = 64, kernel_size = [3, 3], strides = (2, 2), padding = 'same')
in2 = tf.contrib.layers.batch_norm(in2, is_training = is_training)

net = tf.nn.relu(in1) + tf.nn.relu(in2)
```
But this time opencv give me error messages

> strided_slice:StridedSlice(input)(strided_slice/stack)(strided_slice/stack_1)(strided_slice/stack_2)
> shrink_axis_mask:[ ]
> end_mask:[ ]
> new_axis_mask:[ ]
> ellipsis_mask:[ ]
> begin_mask:[ ]
> T:0
> Index:[ ]
> OpenCV Error: Unspecified error (Unknown layer type StridedSlice in op strided_slice) in populateNet, file /home/ramsus/Qt/3rdLibs/opencv/modules/dnn/src/tensorflow/tf_importer.cpp, line 1077
> /home/ramsus/Qt/3rdLibs/opencv/modules/dnn/src/tensorflow/tf_importer.cpp:1077: error: (-2) Unknown layer type StridedSlice in op strided_slice in function populateNet
>

**Comment by stereomatchingkiss** [comment]
Although blobFromImage do not support 2 dim image, we could do it by ourself. 

Following maybe is a correct solution(I haven't checked the output is correct or not), more details are [here](http://answers.opencv.org/question/175425/create-a-two-channels-blob-for-dnnnet/)

```
cv::Mat left_img(128, 128, CV_32F);
cv::Mat right_img(128, 128, CV_32F);
int sz[] = {1, 2, left_img.rows, left_img.cols};
cv::Mat blob = Mat(4, sz, CV_32F);    

cv::Mat blob_0(128, 128, CV_32F, blob.ptr(0,0));
cv::Mat blob_1(128, 128, CV_32F, blob.ptr(0,1));
left_img.copyTo(blob_0);
right_img.copyTo(blob_1);
net.setInput(blob, "input");
Mat result = net.forward();
std::cout<<result.rows<<","<<result.cols<<","<<result.channels()<<std::endl;
```

Is it a good idea to create an api like blobFromNDimImage?

**Comment by dkurt** [comment]
@stereomatchingkiss, `blobFromImage` don't support 2 channel images because we just haven't met a model that processes them for now. Possibly it might be a grayscale with alpha or grayscale with a some depth map.

The issue's topic and it's content are a bit different: merged two single channel images are not 2-channel image. As for me, it's more a **batch** of images what may be created by [blobFromImage**s**](http://docs.opencv.org/master/d6/d0f/group__dnn.html#ga2171d42f5e0c5ee4e39fd5a1acf0837d). Then you'll have a blob of size `2x1x128x128` but I think you need to use a blob of `1x2x128x128`?

By the way, how you merge it in python? NumPy preprocessing or network with two inputs and concatenation layer? In the first way you have to do the same in OpenCV: merge images as mentioned above.

**Comment by stereomatchingkiss** [comment]
> Then you'll have a blob of size 2x1x128x128 but I think you need to use a blob of 1x2x128x128?

Yes, that is why I create it by myself

> By the way, how you merge it in python?

I do not merge it, I stack two gray scale image together. It is an homography estimation model proposed by [Deep image homography](https://arxiv.org/pdf/1606.03798.pdf)

```
img = np.empty(shape=([shape[1],shape[2],shape[3]]), dtype='float32')
img[:,:,0] = misc.imread(folder + '/' + imgs_name[i][0])
img[:,:,1] = misc.imread(folder + '/' + imgs_name[i][1])            
```

**Comment by dkurt** [comment]
So it's neither 2 channels image nor batch of two 1 channel images. That is why we need to prepare a 4d blob manually. Is there an issue here or we can close it?

**Comment by stereomatchingkiss** [comment]
@dkurt Please close it and thanks for your helps
